packages/visibility_detector/lib/src/render_visibility_detector.dart

```diff
@@ -34,33 +98,231 @@ class RenderVisibilityDetector extends RenderProxyBox {
+      // middle of a frame.
+      if (isFirstUpdate) {
+        // We're about to render a frame, so a post-frame callback is guaranteed
+        // to fire and will give us the better immediacy than `scheduleTask<T>`.
+        SchedulerBinding.instance.addPostFrameCallback((timeStamp) {
+          _processCallbacks();
+        });
+      }
+    } else if (_timer == null) {
+      // We use a normal [Timer] instead of a [RestartableTimer] so that changes
+      // to the update duration will be picked up automatically.
+      _timer = Timer(updateInterval, _handleTimer);
+    } else {
+      assert(_timer!.isActive);
+    }
+  }
+
+  VisibilityInfo _determineVisibility(ContainerLayer? layer, Rect bounds) {
+    if (_disposed || layer?.attached == false || !attached) {
+      // layer is detached and thus invisible.
+      return VisibilityInfo(
+        key: key,
+        size: _lastVisibility[key]?.size ?? Size.zero,
+      );
+    }
+    final transform = Matrix4.identity();
+
+    // Create a list of RenderObjects from this to the root, excluding the root
+    // since that has the DPR transform and we want to work with logical pixels.
+    // Cannot use the layer tree since some ancestor render object may have
+    // directly transformed/clipped the canvas. If there is some way to figure
+    // out how to get the RenderObjects below [layer], could take advantage of
+    // the usually shallower height of the layer tree compared to the render
+    // tree. Alternatively, if the canvas itself exposed the current matrix/clip
+    // we could use that.
+    RenderObject? ancestor = parent as RenderObject?;
+
+    final List<RenderObject> ancestors = <RenderObject>[];
+    ancestors.add(this);
+    RenderObject child = this;
+    while (ancestor != null && ancestor.parent != null) {
+      if (!ancestor.paintsChild(child)) {
+        return VisibilityInfo(key: key, size: bounds.size);
+      }
+      ancestors.add(ancestor);
+      child = ancestor;
+      ancestor = ancestor.parent as RenderObject?;
+    }
+
+    // Determine the transform and clip from first child of root down to
+    // this.
+    Rect clip = Rect.largest;
+    for (int index = ancestors.length - 1; index > 0; index -= 1) {
+      final parent = ancestors[index];
+      final child = ancestors[index - 1];
+      Rect? parentClip = parent.describeApproximatePaintClip(child);
+      if (parentClip != null) {
+        clip = clip.intersect(MatrixUtils.transformRect(transform, parentClip));
+      }
+      parent.applyPaintTransform(child, transform);
+    }
+    return VisibilityInfo.fromRects(
+      key: key,
+      widgetBounds: MatrixUtils.transformRect(transform, bounds),
+      clipRect: clip,
+    );
   }
 
-  // See [RenderObject.alwaysNeedsCompositing].
+  bool _disposed = false;
   @override
-  bool get alwaysNeedsCompositing => onVisibilityChanged != null;
+  void dispose() {
+    _compositionCallbackCanceller?.call();
```

Comment:
Do we need to "forget" this RO in dispose so the visibility logic is no longer triggered for it?

Reply:
Yes, that may be our only chance to let clients know that the RO is no longer visible (assuming it ever was). There are tests that fail if you add a forget here.